File: Algorithms/Python/0042_Trapping_Rain_Water/Trapping_Rain_Water_Solution_1.py

# Space: O(1)
# Time: O(n)

class Solution:
    def trap(self, height) -> int:
        length = len(height)
        if length <= 2: return 0

        # find the highest_bar
        highest_bar = 0
        highest_bar_index = 0
        for i in range(length):
            if height[i] > highest_bar:
                highest_bar = height[i]
                highest_bar_index = i

        res = 0

        # calculate the water from left to highest_bar
        pointer_from_left = 0
        highest_bar_from_left = 0
        while pointer_from_left != highest_bar_index:
            highest_bar_from_left = max(highest_bar_from_left, height[pointer_from_left])
            res += highest_bar_from_left - height[pointer_from_left]
            pointer_from_left += 1

        # calculate the water from right to highest_bar
        pointer_from_right = length - 1
        highest_bar_from_right = 0
        while pointer_from_right != highest_bar_index:
            highest_bar_from_right = max(highest_bar_from_right, height[pointer_from_right])
            res += highest_bar_from_right - height[pointer_from_right]
            pointer_from_right -= 1

        return res

File: Algorithms/0042_Trapping_Rain_Water/Java/TrappingRainWaterSolution1.java

// Space: O(1)
// Time: O(n)


class Solution {
    public int trap(int[] height) {
        if (height.length <= 1) return 0;
        int highestBarIndex = 0;
        for (int i = 0; i < height.length; i++) {
            highestBarIndex = height[highestBarIndex] < height[i] ? i : highestBarIndex;
        }
        int res = 0;
        int index = 0;
        int highestBar = 0;
        while (index != highestBarIndex) {
            highestBar = Math.max(highestBar, height[index]);
            res += highestBar - height[index];
            index++;
        }

        index = height.length - 1;
        highestBar = 0;
        while (index != highestBarIndex) {
            highestBar = Math.max(highestBar, height[index]);
            res += highestBar - height[index];
            index--;
        }
        return res;
    }
}




